subiquity/server/controllers/tests/test_desktop.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2022 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import unittest
+from unittest.mock import patch
+
+from subiquity.server.controllers.desktop import (
+    DesktopController,
+    run_as_command
+)
+from subiquitycore.tests.mocks import make_app
+
+
+class TestDesktopController(unittest.IsolatedAsyncioTestCase):
+    def setUp(self):
+        self.app = make_app()
+        self.controller = DesktopController(self.app)
+
+    def test_serialize(self):
+        self.controller.model.username = "foo"
+        self.assertEqual(self.controller.serialize(), "foo")
+
+    def test_deserialize(self):
+        self.controller.deserialize("baz")
+        self.assertEqual(self.controller.model.username, "baz")
+
+    @patch("subiquity.server.controllers.desktop.arun_command")
+    async def test_set_input_source(self, mock_arun_command):
+        self.app.opts.dry_run = False
+        self.controller.model.username = "bar"
+        await self.controller.set_input_source("fr", "latin9")
+        gsettings = [
+            "gsettings", "set", "org.gnome.desktop.input-sources", "sources",
+            "[('xkb','fr+latin9')]"
```

Comment:
A check on a version without a variant would be nice.
Some parts of the subiquity code use `parameterized` to build a list of input data, such as https://github.com/canonical/subiquity/blob/d6cbbda68503e3c1642288f292269f4763d50e78/subiquitycore/tests/test_async_helpers.py#L29

Reply:
the combination of `@parameterized.expand()` and `@patch()` doesn't seem to work with async functions. a simple test case to illustrate the problem:

```py
import os
import unittest

from parameterized import parameterized
from unittest.mock import patch


class TestParameterizedMock(unittest.IsolatedAsyncioTestCase):

    @parameterized.expand([(0,)])
    @patch("os.seteuid")
    def test_parameterized_mock(self, uid, mock_seteuid):
        os.seteuid(uid)  # OK: calls the mock
        mock_seteuid.assert_called_once_with(uid)

    @parameterized.expand([(0,)])
    @patch("os.seteuid")
    async def test_parameterized_mock_async(self, uid, mock_seteuid):
        os.seteuid(uid)  # NOT OK: PermissionError: [Errno 1] Operation not permitted
        mock_seteuid.assert_called_once_with(uid)


if __name__ == '__main__':
    unittest.main()
```

Reply:
```
@parameterized.expand([(0,)])
async def test_parameterized_mock_async(self, uid):
    with patch("os.seteuid") as mock_seteuid:
        os.seteuid(uid)  # OK
        mock_seteuid.assert_called_once_with(uid)
```
`with patch` is a nice construct and would help here.